File: typepos.swift
import Foundation

struct OrderLine {
    var unitPrice: Decimal
    var quantity: Int
}

func totalPrice(lines: AnySequence<OrderLine>) -> Decimal {
    return lines.reduce(into: Decimal(0)) { total, line in
        total += line.unitPrice * Decimal(line.quantity)
    }
}

func main() {
    let line: Any = OrderLine(unitPrice: 0.25, quantity: 2)
    let order = [
        // Though pattern matching better in some of these languages.
        line as! OrderLine,
        OrderLine(unitPrice: 1.00, quantity: 1),
    ]
    print(totalPrice(lines: AnySequence(order)))
}

main()

File: TypePos.cs
using System;
using System.Collections.Generic;
using System.Linq;

class OrderLine {
    public decimal UnitPrice;
    public int Quantity;
}

class TypePos {
    // static decimal TotalPrice<Item>(IEnumerable<Item> lines)
    //         where Item : OrderLine {
    static decimal TotalPrice(IEnumerable<OrderLine> lines) {
        return lines.Aggregate(
            0m, (total, line) => total + line.UnitPrice * line.Quantity
        );
    }

    static void Main() {
        var line = (object)new OrderLine {UnitPrice = 0.25m, Quantity = 2};
        var order = new [] {
            line as OrderLine,
            new OrderLine {UnitPrice = 1.00m, Quantity = 1},
        };
        Console.WriteLine(TotalPrice(order));
    }
}
